curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/65ca2d0625aa3a3201067f70.md

```diff
@@ -0,0 +1,148 @@
+---
+id: 65ca2d0625aa3a3201067f70
+title: Step 30
+challengeType: 0
+dashedName: step-30
+---
+
+# --description--
+
+In `.forEach()`'s callback, use the `el` parameter to access the `counts` object. If the element is not in `counts`, set it to 0. Otherwise, increment the count for the element. Log out the final result of `counts`.
+
+# --hints--
+
+Your .forEach() method should increment the count for each element. Don't forget the initialization scenario.
+
+```js
+/array\.forEach\(\s*(\(\s*el\s*\)|el)\s*=>\s*\{?\s*(if\s*\(.*?\)\s*\{[^}]*\}\s*else\s*)?(counts\[\s*el\s*\]\s*=\s*(counts\[\s*el\s*\]\s*\|\|\s*0)\s*\+\s*1|counts\[\s*el\s*\]\s*=\s*counts\[\s*el\s*\]\s*\+\s*1);?\s*\}?/
```

Comment:
we don't want to use a regex here because that was part of the original problem.
Here is a quote from the original issue

> I would suggest mocking the data so the camper can access it. That should also let us test the result of the forEach without using a regex. There is not only one valid way to solve this making the last regex test questionable.

Even though two possible solutions were mentioned in the issue, there are technically more. And trying to account for all of that using regex is going to be tough to maintain long term.

We should go with the suggestion left on the issue here 

> The test should work the same way, on the actual data and test the expected final object.

We should look into using different assert methods to test the final `counts` object.
That way the camper can write any valid solution and our tests will account for that 
https://www.chaijs.com/api/assert/

If you need help, you can reach out the original poster on the issue for ideas or on the contributors channels on the forum or discord. 👍

Reply:
Okay, I'll look into the Chai Assertion Library

Reply:
The parameter of `getMode()`, `numbers` is defined by the user so how to test the final `counts` object? Do I need to tell the user to enter a specific `numbers`?

Reply:
In step 28, you have them add the console statement for getMode. You can add to the directions that they can test it out with the numbers  `4`, `4`, `2`, `5` and click on the `Calculate` button. 

But in your tests you can probably just hardcode the numbers array of 4,4,2,5 and then use that in your test.

```md
//hint text goes here

const numbersArr = [4,4,2,5]
// test for counts objects goes using the numebrsArr variable
```

Reply:
I am doing this
<img width="565" alt="image" src="https://github.com/freeCodeCamp/freeCodeCamp/assets/62249815/b7682f7d-a7c0-4b1f-badc-8fa5ee0733f2">
but the console says `[ReferenceError: counts is not defined]`

Reply:
I updated the step 30, please take a look